enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,617 @@
+6. Cluster admin edits/creates the CR of the Ingress Controller having Spec as follows -
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+   httpHeaders:
+     headersManipulation:
+       response:
+       - name: X-Frame-Options
+         action: Delete
+         delete:  
+       - name: X-XSS-Protection
+         action: Delete
+         delete:
+```
+7. Users want to delete HTTP response header which need to be sent only to the client hitting
+   the route with the HTTP Response Header values.
+8. Users edits/creates the Route having Spec as follows -
+```yaml
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  labels:
+    type: sharded 
+  name: hello-openshift-edge
+  namespace: hello-openshift
+spec:
+  subdomain: hello-openshift 
+  tls:
+    termination: edge
+  to:
+    kind: Service
+    name: hello-openshift
+    httpHeaders:
+  headersManipulation:
+    response:
+    - name: X-Frame-Options
+      action: Delete
+      delete:
+    - name: X-XSS-Protection
+      action: Delete
+      delete:
+```
+#### Variation [optional]
+
+### API Extensions
+
+A) Set(Replace)/Append/Delete HTTP headers in the global section of haproxy.config via IngressController's Spec.
+The existing Ingress Controller API `HTTPHeaders *IngressControllerHTTPHeaders` is extended by adding an field for action to be performed,
+and field for headers to be set/replaced/appended/deleted.
+
+```go
+// IngressControllerSpec is the specification of the desired behavior of the
+// IngressController.
+type IngressControllerSpec struct {
+    // httpHeaders defines policy for HTTP headers.
+    //
+    // If this field is empty, the default values are used.
+    //
+    // +optional
+    HTTPHeaders *IngressControllerHTTPHeaders `json:"httpHeaders,omitempty"`
+}
+```
+
+```go
+
+// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
+type IngressControllerHeadersManipulation struct {
+    Response []IngressControllerSetReplaceAppendDeleteHTTPHeaders `json:"response,omitempty"`
+}
+
+```
+
+```go
+// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
+// +union
+type IngressControllerSetReplaceAppendDeleteHTTPHeaders struct {
+    // action lets you specify headers for Set or Delete.
+    // +unionDiscriminator
+    // +kubebuilder:validation:Enum:=Set;Delete 
+	//+kubebuilder:validation:Required
+    Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`
+
+    // name specifies a header name.  Its value must be a valid HTTP header
+    // name as defined in RFC 2616 section 4.2.
+    // +kubebuilder:validation:Optional
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=256
+    // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
+    // +optional
```

Comment:
Can empty union member be required? Because we need it required.

Reply:
`Name` should not be a union member.  You can use a inline field for the union type.  See `IngressControllerCaptureHTTPCookie` ([source](https://github.com/openshift/api/blob/5ad311dde2264b659a256a17c27f1836fb50e85d/operator/v1/types_ingress.go#L1100-L1117)) for an example.

Reply:
do you mean something like this -
`    // headersManipulation specifies options for set/replace/append/delete headers.
   HeadersManipulation IngressControllerHeadersManipulation `json:"headersManipulation,omitempty"`
}

type IngressControllerHeadersManipulation struct{
   Response IngressControllerSetOrDeleteHeaders `json:"response,omitempty"`
}

// IngressControllerHeadersManipulation specifies options for set/replace/append/delete headers.
type IngressControllerSetOrDeleteHeaders struct {
   // name specifies a header name to be set(replace)/append/delete.  Its value must be a valid HTTP header.
   // name as defined in RFC 2616 section 4.2.
   // +kubebuilder:validation:Required
   // +kubebuilder:validation:MinLength=1
   // +kubebuilder:validation:MaxLength=256
   // +kubebuilder:validation:Pattern="^[-!#$%&'*+.0-9A-Z^_`a-z|~]+$"
   // +required
   Name *string `json:"name"`

   IngressControllerSetReplaceAppendDeleteHTTPHeaders `json:",inline"`
}

// IngressControllerSetReplaceAppendDeleteHTTPHeaders specifies fields to mention the headers and the action you want to perform on them.
// +union
type IngressControllerSetReplaceAppendDeleteHTTPHeaders struct {
   // action lets you specify headers for Set or Delete.
   // +unionDiscriminator
   // +kubebuilder:validation:Enum:=Set;Delete
   // +kubebuilder:validation:Required
   Action IngressControllerSetHTTPHeaderAction `json:"action,omitempty"`

   // Set defines HTTP headers that should be set or replaced or appended.
   // If this field is empty, no headers are added.
   //
   // Note that this option only applies to cleartext HTTP connections
   // and to secure HTTP connections for which the ingress controller
   // terminates encryption (that is, edge-terminated or reencrypt
   // connections).  Headers cannot be set(replace)/append for TLS passthrough
   // connections.
   //
   // +kubebuilder:validation:Optional
   // +optional
   Set *IngressControllerSetHTTPHeader `json:"set,omitempty"`

   // delete defines HTTP headers that should be deleted.
   // If this field is empty, no headers are deleted.
   //
   // Note that this option only applies to cleartext HTTP connections
   // and to secure HTTP connections for which the ingress controller
   // terminates encryption (that is, edge-terminated or reencrypt
   // connections).  Headers cannot be set(replace)/append for TLS passthrough
   // connections.
   //
   // +kubebuilder:validation:Optional
   // +optional
   Delete *IngressControllerDeleteHTTPHeader `json:"delete,omitempty"`
}`

Reply:
The CR will look like this ->
~~~
apiVersion: operator.openshift.io/v1
kind: IngressController
metadata:
  name: default
  namespace: openshift-ingress-operator
spec:
  httpHeaders:
    headersManipulation:
      response: 
      - name: X-Frame-Options
        action: Set
        set:
          value: DENY
          action: Replace
      - name: X-XSS-Protection
        action: Set
        set:
          value: 1;mode=block
          action: Replace
~~~

Reply:
@Miciah Can you please check this ?

Reply:
Sorry for not responding to this comment earlier.  Do you still need me to check something?

Reply:
api changes are updated